### PR TITLE
feature: Use core CalendarState interface for Android Compose UI

### LINKED ISSUE #5: Android Compose UI
- [x] Create module
- [x] Load data from CalendarPageLoader
- [x] Implement core CalendarState
- [x] Animate CalendarPageLoader changes
- [x] Animate height changes when scrolling
- [x] Snap scrolling

### PR DESCRIPTION
**Pull request by boswelja**
* Fixed a bug where the page that was emitted was incorrect
* EphemerisCalendar's state now implements CalendarState, providing a standardized interface
* Bumped AGP
* Closes #5

### CODE REVIEW

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendar.kt [hunk 1]
```diff
@@ -9,36 +9,21 @@ import androidx.compose.foundation.layout.Row
...
 import androidx.compose.ui.Modifier
 import com.boswelja.ephemeris.core.model.CalendarDay
 import com.boswelja.ephemeris.core.model.CalendarPage
-import com.boswelja.ephemeris.core.ui.CalendarPageLoader
 
 @OptIn(ExperimentalAnimationApi::class)
 @Composable
 public fun EphemerisCalendar(
-    calendarState: CalendarState,
+    calendarState: CalendarStateImpl,
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendar.kt:19]: Passing the implementation explicitly instead of the interface?

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt [hunk 2]
```diff
@@ -21,39 +23,45 @@ import androidx.compose.ui.unit.LayoutDirection
...
+            }
     }
 
     var pagerHeight by remember { mutableStateOf(0) }
 
     // Switch the Modifier used to animate height once pagerHeight has had it's initial update
     val layoutModifier = if (pagerHeight > 0) {
-        Modifier.animateContentSize().then(modifier)
+        Modifier
+            .animateContentSize()
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:58]: this is so unfair...

Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:58]: I'm not super happy with having to switch Modifiers based on the pager height, but `animateContentSize` is terrific
Just don't look down 😆

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt [hunk 3]
```diff
@@ -42,15 +49,33 @@ class ComposeCalendarFragment : Fragment() {
...
+        Text(
+            text = headerText,
+            style = MaterialTheme.typography.titleLarge,
+            modifier = Modifier
+                .padding(horizontal = 16.dp)
+                .padding(top = 24.dp)
+        )
         EphemerisCalendar(
-            modifier = Modifier.semantics { contentDescription = "calendar" },
+            modifier = Modifier.padding(horizontal = 16.dp),
```
[the remark this answers is not in the document]
Comment (boswelja) [reply android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:78]: As soon as you create an issue and start work on it
Remember it's still just a sample

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendar.kt (hunk 1 above)
Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendar.kt:19]: Yep, a couple of reasons
1. Interface doesn't allow for anything other than public, and we don't want to expose all of the inner states. Abstract classes would solve this though
2. I feel like internal states should be implementation specific (pagerState for example), though if you disagree I can abstract a class for this

Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendar.kt]: Have created an abstract class

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt [hunk 4]
```diff
@@ -87,43 +90,46 @@ public fun InfiniteHorizontalPager(
...
-        return internalPage - internalStartPage
+    suspend fun animateScrollToPage(page: Int) {
+        lazyListState.animateScrollToItem(page + internalPageOffset)
     }
 }
 
 @Composable
-public fun rememberInfinitePagerState(startPage: Int = 0): InfinitePagerState {
+internal fun rememberInfinitePagerState(
+    startPage: Int = 0
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:127]: this is internal now, so there's no way for a user to start on a page that isn't 0?

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt [hunk 5]
```diff
@@ -42,15 +50,33 @@ class ComposeCalendarFragment : Fragment() {
...
         CalendarMonthPageSource(
             DayOfWeek.SUNDAY
         )
     }
+
+    var headerText by remember { mutableStateOf("") }
+
+    LaunchedEffect(calendarState) {
+        calendarState.displayedDateRange.collectLatest {
+            headerText = "${it.start.toJavaLocalDate().format(headerDateFormatter)} - ${it.endInclusive.toJavaLocalDate().format(headerDateFormatter)}"
```
Comment (Iannnr) [inline android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:66]: >long line of code
tempted to say extension function here instead? Or at least some line breaks 😅

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                pageLoader = CalendarPageLoader(
+                    coroutineScope,
+                    value
+                )
+            }
+        }
+
+    override fun scrollToDate(date: LocalDate) {
+        // We launch this in a coroutine to avoid any thread blocking
+        coroutineScope.launch {
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:76]: might need to check that the scope is active before trying this, especially as it can be any scope, not specifically a `rememberCoroutineScope`

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    calendarPageSource: CalendarPageSource,
+    private val coroutineScope: CoroutineScope,
+    override val pagerState: InfinitePagerState
+) : EphemerisCalendarState() {
+
+    override var pageLoader by mutableStateOf(CalendarPageLoader(coroutineScope, calendarPageSource))
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override val displayedDateRange: StateFlow<ClosedRange<LocalDate>> = snapshotFlow { pageLoader }
+        .flatMapLatest { snapshotFlow { pagerState.page } }
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:55]: _snapshotFlow entered the chat_

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt [hunk 8]
```diff
@@ -13,57 +14,59 @@ import androidx.compose.runtime.getValue
...
-        Modifier.animateContentSize().then(modifier)
+        Modifier
+            .animateContentSize()
+            .then(modifier)
     } else modifier
 
     LazyRow(
         modifier = layoutModifier,
-        state = internalLazyState,
+        state = state.lazyListState,
```
Comment (Iannnr) [inline android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:60]: reading this made me wonder if it was setting its own state because the variable names clash between `LazyRow#state` and `state: InfinitePagerState = rememberInfinitePagerState()` 😩

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt (hunk 4 above)
Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:127]: In the context of Ephemeris calendar Composable, the start page is dictated by the page source rather than the page number. If users wanted to implement InfoniteHorizontalPager then that's another story :P

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt (hunk 6 above)
Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:76]: What would the expected behavior be if the scope isn't active? Or would you rather not allow users to pass a scope?

Comment (Iannnr) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:76]: it'd be a no-op if it's cancelled, I don't know about `rememberCoroutineScope`, but if it's tied to the lifecycle of the Composable, it's a safer way to ensure that it'll be launched without any of the aforementioned side effects and the user won't see anything happen

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt (hunk 4 above)
Comment (Iannnr) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:127]: So this is just a page 0 to enable the InfinitePager to work, and not actually a calendar page?

Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:127]: Yep, pretty much

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt [hunk 9]
```diff
@@ -42,15 +50,33 @@ class ComposeCalendarFragment : Fragment() {
...
+fun CalendarScreen(
+    headerDateFormatter: DateTimeFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)
+) {
     val calendarState = rememberCalendarState {
         CalendarMonthPageSource(
             DayOfWeek.SUNDAY
         )
     }
+
+    var headerText by remember { mutableStateOf("") }
```
Comment (Iannnr) [inline android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:62]: I heard you liked scope creep, so what if I suggested something like an `onDateChanged` to update this state instead of just a mutable remember state? :P

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt (hunk 4 above)
Comment (Iannnr) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/InfiniteHorizontalPager.kt:127]: Cool, will resolve with no fix planned then

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt (hunk 5 above)
Comment (boswelja) [reply android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:66]: An extension makes sense, we can share it with XML views that way

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt (hunk 7 above)
Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:55]: Pretty much lol

File: android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt (hunk 6 above)
Comment (boswelja) [reply android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt:76]: Yep, `rememberCoroutineScope` is tied to the lifecycle of the call site. I think not allowing users to pass a coroutine scope will work fine in this case

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt (hunk 9 above)
Comment (boswelja) [reply android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:62]: As part of `EphemerisCalendar`? That would have to be a `ClosedRange<LocalDate>`, I like the idea but it does also deviate from the core CalendarState interface

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt (hunk 5 above)
Comment (boswelja) [reply android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:66]: I've created an issue #52 to update sample code

File: android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt (hunk 9 above)
Comment (boswelja) [reply android/sample/src/main/java/com/boswelja/ephemeris/sample/compose/ComposeCalendarFragment.kt:62]: If you'd like to open an issue we can look into this in more depth
